include/deal.II/lac/petsc_vector_base.h

```diff
@@ -912,8 +921,6 @@ namespace PETScWrappers
                (vector.last_action == VectorOperation::unknown),
              ExcWrongMode(VectorOperation::insert, vector.last_action));
 
-      Assert(!vector.has_ghost_elements(), ExcGhostsPresent());
-
```

Comment:
I don't think this is what we want.

The mental model in deal.II is that vectors that have ghost elements are read-only. I know that both Trilinos and PETSc can handle modifying vectors that have ghost elements, but the reality is that that is a prescription for very difficult to find bugs because what you get to see on different processes is no longer guaranteed to be the same. In other words, the fact that we treat ghosted vectors a read-only was a conscious decision meant to make it easier to write correct programs, and I'm not ready to give that invariant up without further discussion.

This assertion, as well as others in this file that you're removing, simply enforces this invariant.

Reply:
Judging from the code, I can tell you that now all the read accesses on the ghosted data will be correct. I made sure of that, by updating ghost values at each compress call.

You need some sort of semantics barrier to enforce those kinds of sanity checks. As it is implemented in this PR, every deal.II interface call that modifies vector values now calls either `update_ghost_values` or `compress`. Also, note that lower-level API modifications go through `VecSetValues`, which already puts the vector in an undefined state until `VecAssemblyBegin/End` is called. 

These changes also go in the direction of a unification of the various distributed vector implementations as discussed here https://github.com/dealii/dealii/issues/14426

Reply:
I don't doubt that the implementation works correctly -- the question is whether we want to allow for the use case you enable here, and I'm not convinced of that.

This might have to be a broader discussion with the other developers as well. Would you be amenable to putting the assertions back in and leaving their removal for a separate patch? This way we could at least get the functional changes in this patch merged!

Reply:
IIRC the check deleted here is very strict - it prevents doing `v = 0;` if `v` has ghost elements. There is no such restriction, AFAICT, for our own `LA::d::V` which has a more fine-grained notion of 'ghosts are present' than 'this vector has ghost entries'. I assemble right-hand sides into ghosted `LA::d::V`s - surely that should be legal for all kinds of vectors (without resorting to things like `VecStash`).

We need to be careful with these semantics but I am open to relaxing "ghosted vectors are read-only and can only be set by assignment to a non-ghosted vector".

Reply:
Yes, `LA::d::V` is the exception, and one that I'm not happy about :-(

Reply:
I will put assertions back